staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/decode_test.go

```diff
@@ -104,7 +106,220 @@ func TestDecode(t *testing.T) {
+			modes:         []cbor.DecMode{modes.Decode},
+			in:            hex("C48221196AB3"), // 4([-2, 27315])
+			want:          []interface{}{int64(-2), int64(27315)},
+			assertOnError: assertNilError,
+		},
+		{
+			name:          "tag 5 Big floats",
+			modes:         []cbor.DecMode{modes.Decode},
+			in:            hex("C5822003"), // 5([-1, 3])
+			want:          []interface{}{int64(-1), int64(3)},
+			assertOnError: assertNilError,
+		},
+		{
+			name:          "simple value 20",
+			modes:         []cbor.DecMode{modes.Decode},
+			in:            hex("F4"), // false
+			want:          false,
+			assertOnError: assertNilError,
+		},
+		{
+			name:          "simple value 21",
+			modes:         []cbor.DecMode{modes.Decode},
+			in:            hex("F5"), // true
+			want:          true,
+			assertOnError: assertNilError,
+		},
+		{
+			name:          "simple value 22",
+			modes:         []cbor.DecMode{modes.Decode},
+			in:            hex("F6"), // nil
+			want:          nil,
+			assertOnError: assertNilError,
+		},
+		{
+			name:  "simple value 23",
+			modes: []cbor.DecMode{modes.Decode},
+			in:    hex("F7"), // undefined
+			assertOnError: assertOnConcreteError(func(t *testing.T, e *cbor.UnmarshalTypeError) {
+
+			}),
+			failureReason: "cbor data with simple value 23 should throw an error.",
+		},
+		{
+			name:          "infinity",
+			modes:         []cbor.DecMode{modes.Decode},
+			in:            hex("F97C00"),
+			want:          math.Inf(1),
+			assertOnError: assertNilError,
+		},
+		{
+			name:          "negative infinity",
+			modes:         []cbor.DecMode{modes.Decode},
+			in:            hex("F9FC00"),
+			want:          math.Inf(-1),
+			assertOnError: assertNilError,
+		},
+		{
+			name:          "min float64 value",
+			modes:         []cbor.DecMode{modes.Decode},
+			in:            hex("FB0000000000000001"),
+			want:          math.SmallestNonzeroFloat64,
+			assertOnError: assertNilError,
+		},
+		{
+			name:          "max float64 value",
+			modes:         []cbor.DecMode{modes.Decode},
+			in:            hex("FB7FEFFFFFFFFFFFFF"),
+			want:          math.MaxFloat64,
+			assertOnError: assertNilError,
+		},
+	}
+
+	generateSimpleValueCases := func(start, end, val int) {
+		for simpleValue := start; simpleValue <= end; simpleValue++ { // from simple(0) to simple(19)
+			testcases = append(testcases, testcase{
+				name:  fmt.Sprintf("simple value %d", val),
+				modes: []cbor.DecMode{modes.Decode},
+				in:    hex(fmt.Sprintf("%X", simpleValue)),
+				assertOnError: assertOnConcreteError(func(t *testing.T, e *cbor.UnmarshalTypeError) {
+
+				}),
+				failureReason: fmt.Sprintf("cbor data with simple value %d should throw an error.", val),
+			})
+			val += 1
+		}
+	}
+
+	generateSimpleValueCases(0b11100000, 0b11110011, 0)                  // generates test cases for simple values from 0 to 19
+	generateSimpleValueCases(0b1111100000011000, 0b1111100011111111, 24) // generates test cases for simple values from 24 to 255
```

Comment:
I think the intent would be clearer as a single loop:

```suggestion
// Generate cases for all possible 256 simple values.
for i := 0; i <= 255; i++ {
  each := testcase{
    name: fmt.Sprintf("simple value %d", i),
    want: nil,
  }
  if i <= 23 {
    each.in = []byte{byte(0xe0) | byte(i)}
  } else {
    // larger simple values encode to two bytes
    each.in = []byte{byte(0xe0) | byte(24), byte(i)}
  }

  switch i {
  case 20, 21, 22, 23: // recognized values with explicit cases
    continue
  case 24, 25, 26, 27, 28, 29, 30, 31: // reserved
    // these are considered not well-formed
    each.assertOnError = assertOnConcreteError(func(t *testing.T, e *cbor.SyntaxError) {
      // TODO: ...
    })
  default:
    // don't accept unrecognized values
    each.assertOnError = func(t *testing.T, e error) {
      // TODO: Once implemented, this can be a stronger assertion.
      if e == nil {
        t.Error("expected non-nil error")
      }
    }
    each.failureReason = "unrecognized simple values should be rejected"
  }

  testcases = append(testcases, each)
}
```

Reply:
Fixed.